google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java

```diff
@@ -1255,6 +1260,447 @@ public static SignUrlOption withQueryParams(Map<String, String> queryParams) {
+      this.expires = builder.expires;
+      this.successActionRedirect = builder.successActionRedirect;
+      this.successActionStatus = builder.successActionStatus;
+
+      this.fieldsMap = new HashMap<>();
+      fieldsMap.put("acl", acl);
+      fieldsMap.put("cache-control", cacheControl);
+      fieldsMap.put("content-disposition", contentDisposition);
+      fieldsMap.put("content-length", contentLength);
+      fieldsMap.put("content-encoding", contentEncoding);
+      fieldsMap.put("content-type", contentType);
+      fieldsMap.put("expires", expires);
+      fieldsMap.put("success_action_redirect", successActionRedirect);
+      fieldsMap.put("success_action_status", successActionStatus);
+    }
+
+    private V4PostFields(Map<String, String> fields) {
+      this.fieldsMap = fields;
+    }
+
+    public static V4PostFields of(Map<String, String> fields) {
+      return new V4PostFields(fields);
+    }
+
+    public static Builder newBuilder() {
+      return new Builder();
+    }
+
+    public Map<String, String> getFieldsMap() {
+      return fieldsMap;
+    }
+
+    public static class Builder {
+      private Map<String, String> fieldsMap;
+      private String acl;
+      private String cacheControl;
+      private String contentDisposition;
+      private String contentEncoding;
+      private String contentLength;
+      private String contentType;
+      private String expires;
+      private String successActionRedirect;
+      private String successActionStatus;
+
+      private Builder() {
+        fieldsMap = new HashMap<>();
+      }
+
+      public V4PostFields build() {
+        return new V4PostFields(this);
+      }
+
+      public Builder setAcl(String acl) {
+        this.acl = acl;
+        return this;
+      }
+
+      public Builder setCacheControl(String cacheControl) {
+        this.cacheControl = cacheControl;
+        return this;
+      }
+
+      public Builder setContentDisposition(String contentDisposition) {
+        this.contentDisposition = contentDisposition;
+        return this;
+      }
+
+      public Builder setContentEncoding(String contentEncoding) {
+        this.contentEncoding = contentEncoding;
+        return this;
+      }
+
+      public Builder setContentLength(int contentLength) {
+        this.contentLength = "" + contentLength;
+        return this;
+      }
+
+      public Builder setContentType(String contentType) {
+        this.contentType = contentType;
+        return this;
+      }
+
+      public Builder Expires(String expires) {
+        this.expires = expires;
+        return this;
+      }
+
+      public Builder setSuccessActionRedirect(String successActionRedirect) {
+        this.successActionRedirect = successActionRedirect;
+        return this;
+      }
+
+      public Builder setSuccessActionStatus(int successActionStatus) {
+        this.successActionStatus = "" + successActionStatus;
+        return this;
+      }
+    }
+  }
+
+  class V4PostConditions {
```

Comment:
same as above. This is too many inner classes, and the outer class is over 1000 lines. For sheer size, try breaking this up into independent, outer package protected classes.

Reply:
+1 could you split this up into separate files for maintainability as well?